### PR TITLE
Improve test_coverage for github-detail bot

### LINKED ISSUE #122: Get maintained bots to 100% test coverage.
**UPDATE Scroll down to the latest comments to find a list of the remaining bot files not at 100% coverage.**

**NOTE: All striked-through bots have been deprecated, all checked bots have been given tests. See the comments for more information on how to proceed.**

We have a couple bots in `zulip_bots/zulip_bots/bots` that are untested:

~~- [ ] commute~~
~~- [ ]  foursquare~~
~~- [ ] git_hub_comment~~
~~- [ ] github~~

- [x] github_issues

~~- [ ] googlesearch~~
~~- [ ] howdoi~~
~~- [ ] john~~
~~- [ ] youtube~~

For each bot, we should **either**

* add tests. You can find help in [our guide for bot testing](http://zulip.readthedocs.io/en/latest/bots-guide.html?highlight=bot%20testing#writing-tests-for-bots).

* deprecate the bot. If a bot seems to be bugged or completely useless, we should discuss it on [chat.zulip.org](chat.zulip.org) and decide whether it is worth keeping.

### PR DESCRIPTION
**Pull request by amanagr**
Issue #122

### CODE REVIEW

File: zulip_bots/zulip_bots/bots/github_detail/test_github_detail.py [hunk 1]
```diff
@@ -89,3 +89,11 @@ def test_too_many_request(self) -> None:
 
         with self.mock_config_info(self.mock_config):
             self.verify_reply(request, bot_response)
+
+    def test_no_owner_repo(self) -> None:
```
Comment (roberthoenig) [inline zulip_bots/zulip_bots/bots/github_detail/test_github_detail.py]: Maybe name this "test_owner_and_repo_not_specified".

File: zulip_bots/zulip_bots/bots/github_detail/test_github_detail.py [hunk 2]
```diff
@@ -89,3 +89,11 @@ def test_too_many_request(self) -> None:
...
+        mock_invalid_config = {'owner': '', 'repo': ''}
+        request = '/#1'
+        bot_response = 'Failed to detect owner and repository name.'
+
+        with self.mock_config_info(mock_invalid_config):
+            self.verify_reply(request, bot_response)
```
Comment (roberthoenig) [inline zulip_bots/zulip_bots/bots/github_detail/test_github_detail.py]: Can you add two additional tests for the following cases:
* Owner and repo are specified in the config file, but not in the message.
* Owner and repo are specified in the message, but not in the config file.

File: zulip_bots/zulip_bots/bots/virtual_fs/test_virtual_fs.py [hunk 3]
```diff
@@ -19,6 +20,25 @@ class TestVirtualFsBot(BotTestCase):
                 '```\n'
                 'Use commands like `@mention-bot help write` for more details on specific\ncommands.\n')
 
+    def verify_dialog(self, conversation):
```
Comment (roberthoenig) [inline zulip_bots/zulip_bots/bots/virtual_fs/test_virtual_fs.py]: Is there a reason you can't use the default BotTestCase.verify_dialog method?

Comment (amanagr) [reply zulip_bots/zulip_bots/bots/virtual_fs/test_virtual_fs.py]: I wanted to check the case when there are multiple recipients. 
A better way would be to make a BotTestCase.mock_request_message function that sends multiple recipients.

File: zulip_bots/zulip_bots/bots/github_detail/test_github_detail.py [hunk 4]
```diff
@@ -89,3 +89,17 @@ def test_too_many_request(self) -> None:
...
+        valid_request = 'zulip/zulip#5365'
+        bot_response = 'Failed to detect owner and repository name.'
+
+        with self.mock_config_info(mock_invalid_config):
+            self.verify_reply(invalid_request, bot_response)
+            self.verify_reply(valid_request, self.bot_github_response)
+
+        with self.mock_config_info(self.mock_config):
+            self.verify_reply(invalid_request, self.bot_github_response)
+            self.verify_reply(valid_request, self.bot_github_response)
```
Comment (roberthoenig) [inline zulip_bots/zulip_bots/bots/github_detail/test_github_detail.py]: The tests now seem complete, but the different sub-tests in this function could be hard to follow for someone who sees this code for the first time.  Another thing is that every test here should use mocks for http traffic, instead of relying on network access. So to address both of these things, how about this:
* instead of reusing `bot_response_github`, which obfuscates the code, make new fixtures for the cases "valid request" and "invalid request" (for valid request, you can probably reuse an existing fixture, but you should still copy the response into the test functions).
* split the tests in this function into three functions: 
  * test_owner_and_repo_not_specified
  * test_owner_and_repo_specified_in_config_file
  * test_owner_and_repo_specified_in_message

Comment (amanagr) [reply zulip_bots/zulip_bots/bots/github_detail/test_github_detail.py]: Thanks for your detailed review, I will do the changes. :)

**Comment by timabbott** [comment]
@amanagr it looks like you made some changes but then didn't post to let the reviewer know you'd done them -- this probably needs a next review.

**Comment by amanagr** [comment]
@roberthoenig I think it's ready for review.
Thanks @timabbott :)

File: zulip_bots/zulip_bots/bots/github_detail/test_github_detail.py [hunk 5]
```diff
@@ -89,3 +90,45 @@ def test_too_many_request(self) -> None:
 
         with self.mock_config_info(self.mock_config):
             self.verify_reply(request, bot_response)
+
+    def test_owner_and_repo_not_specified(self) -> None:
+        request = '/#1'
+        bot_response = 'Failed to detect owner and repository name.'
+        with self.mock_config_info(self.empty_config):
+            self.verify_reply(request, bot_response)
```
Comment (roberthoenig) [inline zulip_bots/zulip_bots/bots/github_detail/test_github_detail.py]: This test looks like it still needs a mock_http_info.

File: zulip_bots/zulip_bots/test_lib.py [hunk 6]
```diff
@@ -172,3 +172,14 @@ def mock_http_conversation(self, test_name):
...
+    def mock_request_message(self, mock_message):
+        # type: (Dict[str, object]) -> Any
+        message = dict(
+            display_recipient=mock_message.get('display_recipient', 'foo_stream'),
+            sender_email=mock_message.get('sender_email', 'foo@example.com'),
+            sender_full_name=mock_message.get('sender_full_name', 'Foo Test User'),
+            sender_id=mock_message.get('sender_id', '123'),
+            content=mock_message.get('content', ''),
+        )
+        return patch('zulip_bots.test_lib.BotTestCase.make_request_message', return_value=message)
```
Comment (roberthoenig) [inline zulip_bots/zulip_bots/test_lib.py]: This looks like a big function for something that could be accomplished with one additional line of code in the real test. I recommend to proceed like stated in the description of `make_request_message`:
> This is mostly used internally but
tests can override this behavior by
mocking/subclassing.

So in short, just directly patch the function in the test where you are using it.

File: zulip_bots/zulip_bots/bots/virtual_fs/test_virtual_fs.py [hunk 7]
```diff
@@ -19,6 +19,17 @@ class TestVirtualFsBot(BotTestCase):
                 '```\n'
                 'Use commands like `@mention-bot help write` for more details on specific\ncommands.\n')
 
+    def test_multiple_recipient_conversation(self) -> None:
+        expected = [
+            ("mkdir home", "foo@example.com:\ndirectory created"),
+        ]
+        mock_message = dict(
+            display_recipient=[{'email': 'foo_stream'}, {'email': 'mutli_stream'}],
```
Comment (roberthoenig) [inline zulip_bots/zulip_bots/bots/virtual_fs/test_virtual_fs.py]: multi_stream

File: zulip_bots/zulip_bots/bots/github_detail/test_github_detail.py (hunk 5 above)
Comment (amanagr) [reply zulip_bots/zulip_bots/bots/github_detail/test_github_detail.py]: Since, it doesn't has a `owner or repo` it doen't access the `get_details_from_github` and hence doesn't need a mock_http_info.

**Comment by amanagr** [comment]
Sorry for the weird mistakes, I will try not to repeat them.
Thanks.

File: zulip_bots/zulip_bots/bots/github_detail/test_github_detail.py (hunk 5 above)
Comment (roberthoenig) [reply zulip_bots/zulip_bots/bots/github_detail/test_github_detail.py]: Right, makes sense.

File: zulip_bots/zulip_bots/bots/virtual_fs/test_virtual_fs.py [hunk 8]
```diff
@@ -56,14 +71,29 @@ def test_commands_2(self) -> None:
...
             ("rmdir /home", "foo@example.com:\nremoved"),
+            ("pwd", "foo@example.com:\nERROR: the current directory does not exist"),
             ("ls", "foo@example.com:\nERROR: file does not exist"),
+            ("ls foo/", "foo@example.com:\nERROR: foo/ is not a valid name"),
+            ("rm foo/", "foo@example.com:\nERROR: foo/ is not a valid name"),
+            ("rmdir foo/", "foo@example.com:\nERROR: foo/ is not a valid name"),
+            ("write foo/ a", "foo@example.com:\nERROR: foo/ is not a valid name"),
+            ("read foo/", "foo@example.com:\nERROR: foo/ is not a valid name"),
+            ("rmdir /foo", "foo@example.com:\nERROR: directory does not exist"),
+
```
Comment (roberthoenig) [inline zulip_bots/zulip_bots/bots/virtual_fs/test_virtual_fs.py]: Redundant newline.

File: zulip_bots/zulip_bots/bots/virtual_fs/test_virtual_fs.py [hunk 9]
```diff
@@ -19,6 +20,20 @@ class TestVirtualFsBot(BotTestCase):
                 '```\n'
                 'Use commands like `@mention-bot help write` for more details on specific\ncommands.\n')
 
+    def test_multiple_recipient_conversation(self) -> None:
+        expected = [
+            ("mkdir home", "foo@example.com:\ndirectory created"),
+        ]
+        message = dict(
+            display_recipient=[{'email': 'foo_stream'}, {'email': 'multi_stream'}],
```
Comment (roberthoenig) [inline zulip_bots/zulip_bots/bots/virtual_fs/test_virtual_fs.py]: Thinking about this, there isn't such a thing as "multi_streams" - if display_recipient is a list of multiple recipients, those recipients are all persons. Can you change the email entries accordingly?

File: zulip_bots/zulip_bots/bots/github_detail/test_github_detail.py [hunk 10]
```diff
@@ -89,3 +90,45 @@ def test_too_many_request(self) -> None:
...
+                       ' - [WIP] modal: Replace bootstrap modal with custom modal class**\n'\
+                       'Created by **[jackrzhang](https://github.com/jackrzhang)**\n'\
+                       'Status - **Open**\n```quote\nAn interaction bug (#4811)  '\
+                       'between our settings UI and the bootstrap modals breaks hotkey '\
+                       'support for `Esc` when multiple modals are open.\r\n\r\ntodo:\r\n[x]'\
+                       ' Create `Modal` class in `modal.js` (drafted by @brockwhittaker)\r\n[x]'\
+                       ' Reimplement change_email_modal utilizing `Modal` class\r\n[] Dump '\
+                       'using bootstrap for the account settings modal and all other modals,'\
+                       ' replace with `Modal` class\r\n[] Add hotkey support for closing the'\
+                       ' top modal for `Esc`\r\n\r\nThis should also be a helpful step in removing dependencies from Bootstrap.\n```'
```
Comment (roberthoenig) [inline zulip_bots/zulip_bots/bots/github_detail/test_github_detail.py]: This line is too long.

File: zulip_bots/zulip_bots/bots/github_detail/test_github_detail.py [hunk 11]
```diff
@@ -89,3 +90,45 @@ def test_too_many_request(self) -> None:
...
```
Comment (roberthoenig) [inline zulip_bots/zulip_bots/bots/github_detail/test_github_detail.py]: This line is too long.

**Comment by amanagr** [comment]
@roberthoenig I have update the code according your review.
Thanks.

**Comment by roberthoenig** [comment]
Cool, merging. Thanks @amanagr!
